listenbrainz_spark/stats/user/artist.py

```diff
@@ -0,0 +1,146 @@
+import time
+from collections import defaultdict
+from datetime import datetime
+from flask import current_app
+
+from listenbrainz_spark.constants import LAST_FM_FOUNDING_YEAR
+from listenbrainz_spark.path import LISTENBRAINZ_DATA_DIRECTORY
+from listenbrainz_spark.stats import replace_months, run_query, adjust_days, replace_days
+from listenbrainz_spark.stats.user.utils import filter_listens, get_latest_listen_ts
+from listenbrainz_spark.utils import get_listens
+
+
+def get_artists(table):
+    """ Get artist information (artist_name, artist_msid etc) for every user
+        ordered by listen count in a particular time range.
+
+        Args:
+            table (str): name of the temporary table.
+            from_date (str): UNIX timestamp of start time, defaults to LAST_FM_FOUNDING_YEAR
+            to_date (str): UNIX timestamp of end time, defaults to NOW
+
+        Returns:
+            artists: A dict of dicts which can be depicted as:
+                    {
+                        'user1': [{
+                            'artist_name': str, 'artist_msid': str, 'artist_mbids': str, 'listen_count': int
+                        }],
+                        'user2' : [{...}]
+                    }
+    """
+
+    t0 = time.time()
+
+    result = run_query("""
+            SELECT user_name
+                 , artist_name
+                 , artist_msid
+                 , artist_mbids
+                 , count(artist_name) as cnt
+              FROM {table}
+          GROUP BY user_name
+                 , artist_name
+                 , artist_msid
+                 , artist_mbids
+          ORDER BY cnt DESC
+            """.format(table=table))
+
+    rows = result.collect()
+    artists = defaultdict(list)
+    for row in rows:
+        artists[row.user_name].append({
+            'artist_name': row['artist_name'],
+            'artist_msid': row['artist_msid'],
+            'artist_mbids': row['artist_mbids'],
+            'listen_count': row['cnt'],
+        })
+    print("Query to calculate artist stats processed in %.2f s" % (time.time() - t0))
+    return artists
+
+
+def get_artists_last_week():
+    current_app.logger.debug("Calculating artist_last_week...")
+
+    date = get_latest_listen_ts()
+
+    # Get date for Monday before "date"
+    to_date = adjust_days(date, date.weekday())
```

Comment:
We should move this logic out to a seperate function, that would make it easier to test.